sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala

```diff
@@ -0,0 +1,107 @@
+package scalajsbundler.sbtplugin
+
+import org.scalajs.core.tools.io.FileVirtualJSFile
+import org.scalajs.sbtplugin.ScalaJSPlugin
+import org.scalajs.sbtplugin.ScalaJSPlugin.autoImport._
+import sbt.Keys._
+import sbt._
+
+import scalajsbundler._
+
+object NpmDepsPlugin extends AutoPlugin {
+
+  override lazy val requires = ScalaJSPlugin
+
+  // Exported keys
+  object autoImport {
+
+    /**
+      * Fetches NPM dependencies. Returns the directory in which the `npm install` command has been run.
+      *
+      * The plugin uses different directories according to the configuration (`Compile` or `Test`). Thus,
+      * this setting is scoped by a `Configuration`.
+      *
+      * Typically, if you want to define a task that uses the downloaded NPM packages you should
+      * specify the `Configuration` you are interested in:
+      *
+      * {{{
+      *   myCustomTask := {
+      *     val npmDirectory = (npmUpdate in Compile).value
+      *     doSomething(npmDirectory / "node_modules" / "some-package")
+      *   }
+      * }}}
+      *
+      * The task returns the directory in which the dependencies have been fetched (the directory
+      * that contains the `node_modules` directory).
+      *
+      * @group tasks
+      */
+    val npmUpdate: TaskKey[File] =
+      taskKey[File]("Fetch NPM dependencies")
+
+    /**
+      * Whether to use [[https://yarnpkg.com/ Yarn]] to fetch dependencies instead
+      * of `npm`. Yarn has a caching mechanism that makes the process faster.
+      *
+      * If set to `true`, it requires Yarn 0.22.0+ to be available on the
+      * host platform.
+      *
+      * Defaults to `false`.
+      *
+      * @group settings
+      */
+    val useYarn: SettingKey[Boolean] =
+      settingKey[Boolean]("Whether to use yarn for updates")
+  }
+
+  import autoImport._
+
+  val scalaJSBundlerPackageJson =
+    TaskKey[BundlerFile.PackageJson]("scalaJSBundlerPackageJson",
+      "Write a package.json file defining the NPM dependencies of project",
+      KeyRanks.Invisible
+    )
+
+  override lazy val projectSettings =
+    perScalaJSStageSettings(fullOptJS) ++ perScalaJSStageSettings(fastOptJS)
+
+  protected def perScalaJSStageSettings(stage: TaskKey[Attributed[File]]): Seq[Def.Setting[_]] = Seq(
+    useYarn := false
+  ) ++ inConfig(Compile)(perConfigSettings)
+
+
+  private lazy val perConfigSettings: Seq[Def.Setting[_]] = Seq(
+    npmUpdate := {
+      val log = streams.value.log
+      val targetDir = (crossTarget in npmUpdate).value
+      val jsResources = scalaJSNativeLibraries.value.data
+      val packageJsonFile = scalaJSBundlerPackageJson.value
```

Comment:
You're using `scalaJSBundlerPackageJson.value` from `NpmDepsPlugin`, but `NpmDepsPlugin` does not assign any value to that task. I'm not sure how `NpmDepsPlugin` can be used on its own under those circumstances.

You should probably add a scripted test with a project that only uses `NpmDepsPlugin` (but not `ScalaJSBundlerPlugin`), and does something meaningful with it. You might uncover other similar issues.

Reply:
scalaJSBundlerPackageJson is set in ScalaJSBundlerPlugin ```perConfigSettings``` and set it in my ExecNpmPlugin ```perConfigSettings```

Reply:
That means that `NpmDepsPlugin` is not useful on its own, so it is probably not a good delineation of an abstraction. `NpmDepsPlugin` should define `scalaJSBundlerPackageJson` itself (and it should be called `npmDepsPackageJson`), in a way that is generic enough to be reused.

From the way it is defined in `ScalaJSBundlerPlugin`, I don't see why the definition there cannot be factored out (almost) as is into `NpmDepsPlugin`. The only bundler-specific things there are `(version in webpack).value` and `(version in startWebpackDevServer).value`, which means that those two things need an additional hook in `NpmDepsPlugin`. Other keys such as `npmDependencies` can be extracted into `NpmDepsPlugin` as well. It makes sense. Then in your plugin `ExecNpmPlugin` you could automatically derive `npmDependencies` from your `allNpmDeps`, and leave the other ones alone since they would have sensible defaults. Doing it this way will result in even less code in `ExecNpmPlugin`, which kind of proves that it would provide a better abstraction.

Reply:
I didn't define ```scalaJSBundlerPackageJson``` in NpmDepsPlugin because I don't want it to depend on ```webpack```. I redefine ```scalaJSBundlerPackageJson``` with fake values in my ```ExecNpmPlugin```. So, what should I do ? Define two new settingsKey for both versions at the ```NpmDepsPlugin``` level (so that their values are set in ```scalaJSBundlerPlugin``` with the webpack version values and with fake ones in the ```ExecNpmPlugin```) ?

Reply:
> Define two new settingsKey for both versions at the NpmDepsPlugin level (so that their values are set in scalaJSBundlerPlugin with the webpack version values and with fake ones in the ExecNpmPlugin) ?

Yes, I think that's what is needed, here.

Reply:
Or perhaps better, remove `webpackVersion` and `webpackDevServerVersion` from the APIs of `PackageJson` altogether. In that case, `ScalaJSBundlerPlugin` would automatically add the relevant pairs to the `npmDevDependencies`, as is currently done [here](https://github.com/scalacenter/scalajs-bundler/blob/e089cf44e0066e484a8799cc35019ea8a406ff14/sbt-scalajs-bundler/src/main/scala/scalajsbundler/PackageJson.scala#L42-L59).

So `ScalaJSBundlerPlugin.perConfigSettings` would contain something like
```scala
npmDevDependencies ++= {
  val webpackVersion := (version in webpack).value
  val webpackDevServerVersion := (version in startWebpackDevServer).value

   val sourceMapLoaderVersion = 
      NpmPackage(webpackVersion).major match {
        case Some(1) | Some(2) => "0.1.5"
        case Some(3) => "0.2.1"
        case Some(x) => sys.error(s"Unsupported webpack major version $x")
        case None => sys.error("No webpack version defined")
      }

    npmDevDependencies.value ++ Seq(
      "webpack" -> webpackVersion,
      "webpack-dev-server" -> webpackDevServerVersion,
      "concat-with-sourcemaps" -> "1.0.4", // Used by the reload workflow
      "source-map-loader" -> sourceMapLoaderVersion // Used by webpack when emitSourceMaps is enabled
    )
}
```
(more or less)

and then `NpmDepsPlugin` wouldn't have to even know about `webpackVersion` or `webpackDevServerVersion`. And you wouldn't need any fake version for those in your plugin. :)

Reply:
Yes, but I still have to define ```npmDependencies``` and stuff in NpmDepsPlugin, which does not really make sense. 
Actually I don't need to have ```devDependencies``` to be written at all in my package.json. Why not having a ```writePackageJson``` function with optional ```devDependencies``` arguments(```npmDependencies.value```, ```npmDevDependencies.value```, ```npmResolutions.value```,```additionalNpmConfig.value```, and both webpack versions) ?

Reply:
Why would `npmDependencies` not make sense in `NpmDepsPlugin`!? On the contrary, it makes *perfect* sense. It's even the name of the plugin: *NpmDeps*Plugin.

And the other arguments are already "optional" in the sense that they are collections, and their default values are to be empty. No need to make them more optional than that.

Reply:
> Why would npmDependencies not make sense in NpmDepsPlugin!?

Deps is not the same than ```Dependencies```, it is a collection of ``` case class Dep(module: String, version: String, jsFiles: List[String])```. But now it has been moved in ```ExecNpmPlugin```. So ```NpmDepsPlugin``` should be renamed to avoid confusion. Something like NpmUpdatePlugin.

And, you right, adding Options is ugly. 
Well, it all about 3 lines of json to be added or not ! :)

Reply:
I splited the write function of package.json into devDependencies and depencies parts. The ```PackageJson.write``` method is still called from ```ScalaJSBundlerPlugin``` and ```PackageJson.writeDepsOnly``` is called from my ```ExecNpmPlugin```. This way, no more fake values and partial fake package.json file are generated.

Reply:
> Deps is not the same than Dependencies, it is a collection of case class Dep(module: String, version: String, jsFiles: List[String]).

My point all along has been that you can *derive* `npmDependencies` from what your `Dep`s. And that therefore, `NpmDepsPlugin` (which should perhaps be called `NpmDependenciesPlugin`, to avoid naming clashes) can be implemented in terms of `npmDependencies`, and that your plugin can automatically derive `npmDependencies` in terms of what you call `npmDeps`.

If we're not going at least *that* far in factoring out things out of `ScalaJSBundlerPlugin`, then `NpmUpdatePlugin` is really *just* a `def`, and should not be an sbt plugin at all.